### PR TITLE
[CPU] Drop redundant MemoryOutput nodes

### PR DESCRIPTION
**Pull request by maxnick**
### Details:
In direct ReadValue->Assign pairs the Assign node is practically useless as there are no other layers that might modify data in between. Therefore, it does make sense to remove corresponding MemoryOutput nodes to eliminate additional overheads on their processing.

### Tickets:
 - CVS-153035
 - CVS-155112

### CODE REVIEW

File: src/tests/functional/shared_test_classes/src/subgraph/lora_pattern.cpp [hunk 1]
```diff
@@ -37,6 +38,78 @@ void LoraPatternBase::run_test_empty_tensors() {
...
+
+        auto tx_result = inferRequest.get_tensor(outputs[0]);
+        auto tz_result = inferRequest.get_tensor(outputs[1]);
+
+        auto tx_result_ref = inferRequestRef.get_tensor(outputs[0]);
+        auto tz_result_ref = inferRequestRef.get_tensor(outputs[1]);
+
+        ov::test::utils::compare(tx_result, tx_result_ref, 1e-4, 1e-4);
+        ov::test::utils::compare(tz_result, tz_result_ref, 1e-4, 1e-4);
+    }
```
Comment (xipingyan) [inline src/tests/functional/shared_test_classes/src/subgraph/lora_pattern.cpp:110]: Maybe we also need to add check if exist MemoryOutput in compiledModel.

File: src/plugins/intel_cpu/src/graph_optimizer.cpp [hunk 2]
```diff
@@ -3186,5 +3190,94 @@ void GraphOptimizer::MatchSdpaKvCache(Graph &graph) {
...
+        for (auto&& item : childEdges) {
+            auto childNode = item->getChild();
+
+            if (Type::MemoryOutput == childNode->getType()) {
+                if (MemoryOutput && MemoryOutput != childNode) {
+                    //only one child MemoryOutput is expected
+                    return false;
+                }
+                MemoryOutput = childNode;
+            }
```
Comment (xipingyan) [inline src/plugins/intel_cpu/src/graph_optimizer.cpp:3215]: Is it possible there is a case that ReadValue follows 2 Assginns, only 1 Assgin have same variable ID with current ReadValue?
If yes, I suggest only check if exist ReadValue follow Assgin with same Variable Id. What's your idea?

File: src/tests/functional/shared_test_classes/src/subgraph/lora_pattern.cpp (hunk 1 above)
Comment (maxnick) [reply src/tests/functional/shared_test_classes/src/subgraph/lora_pattern.cpp:110]: Yes, for sure, but this is a common test, therefore we cannot add plugin specific checks there. To test the topology we have to introduce a dedicated CPU SL test, and I'm going to do this soon.

File: src/plugins/intel_cpu/src/graph_optimizer.cpp (hunk 2 above)
Comment (maxnick) [reply src/plugins/intel_cpu/src/graph_optimizer.cpp:3215]: Probably it does make sense, you're right. I will extend the condition.

Comment (maxnick) [reply src/plugins/intel_cpu/src/graph_optimizer.cpp]: Done

File: src/plugins/template/src/variable_state.hpp [hunk 3]
```diff
@@ -7,22 +7,28 @@
...
+        : ov::IVariableState(variable_info.variable_id),
+          m_data_shape(variable_info.data_shape),
+          m_data_type(variable_info.data_type),
           m_variable_value(variable_value) {
         m_state = get_tensor_impl(variable_value->get_state());
     }
     void set_state(const ov::SoPtr<ov::ITensor>& state) override {
-        OPENVINO_ASSERT(state->get_shape() == m_state->get_shape(), "Wrong tensor shape.");
-        OPENVINO_ASSERT(state->get_element_type() == m_state->get_element_type(), "Wrong tensor type.");
+        OPENVINO_ASSERT(m_data_shape.compatible(state->get_shape()), "Wrong tensor shape");
```
Comment (itikhono) [inline src/plugins/template/src/variable_state.hpp:26]: minor: it's better to add ore details in the message,
 e.g.
a new state->get_shape() shape have to be compatible with m_data_shape variable shape.

**Review by itikhono — APPROVED** [review]
the changes in the template plugin LGTM

File: src/plugins/template/src/variable_state.hpp (hunk 3 above)
Comment (maxnick) [reply src/plugins/template/src/variable_state.hpp]: Done

**Comment by maxnick** [comment]
@xipingyan , do you have any further comments?

**Review by xipingyan — APPROVED** [review]
LGTM

**Review by praasz — APPROVED** [review]

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**